source/common/filter/config_discovery_impl.h

```diff
@@ -223,6 +223,51 @@ class HttpDynamicFilterConfigProviderImpl
   FactoryCtx& factory_context_;
 };
 
+// Struct of canonical filter name and network filter factory callback.
+struct NamedNetworkFilterFactoryCb {
+  // Canonical filter name.
+  std::string name;
+  // Factory function used to create filter instances.
+  Network::FilterFactoryCb factory_cb;
+};
+
+template <class FactoryCtx, class NeutralNetworkFilterConfigFactory>
+class NetworkDynamicFilterConfigProviderImpl
+    : public DynamicFilterConfigProviderImpl<NamedNetworkFilterFactoryCb> {
+public:
+  NetworkDynamicFilterConfigProviderImpl(
+      FilterConfigSubscriptionSharedPtr& subscription,
+      const absl::flat_hash_set<std::string>& require_type_urls,
+      Server::Configuration::ServerFactoryContext& server_context, FactoryCtx& factory_context,
+      ProtobufTypes::MessagePtr&& default_config, bool last_filter_in_filter_chain,
+      const std::string& filter_chain_type, absl::string_view stat_prefix,
+      const Network::ListenerFilterMatcherSharedPtr& listener_filter_matcher)
+      : DynamicFilterConfigProviderImpl(subscription, require_type_urls,
+                                        server_context.threadLocal(), std::move(default_config),
+                                        last_filter_in_filter_chain, filter_chain_type, stat_prefix,
+                                        listener_filter_matcher),
+        server_context_(server_context), factory_context_(factory_context) {}
+  void validateMessage(const std::string& config_name, const Protobuf::Message& message,
+                       const std::string& factory_name) const override {
+    auto* factory =
+        Registry::FactoryRegistry<NeutralNetworkFilterConfigFactory>::getFactory(factory_name);
+    const bool is_terminal_filter = factory->isTerminalFilterByProto(message, server_context_);
+    Config::Utility::validateTerminalFilters(config_name, factory_name, filter_chain_type_,
+                                             is_terminal_filter, last_filter_in_filter_chain_);
+  }
+
+private:
+  NamedNetworkFilterFactoryCb
+  instantiateFilterFactory(const Protobuf::Message& message) const override {
+    auto* factory = Registry::FactoryRegistry<NeutralNetworkFilterConfigFactory>::getFactoryByType(
+        message.GetTypeName());
+    return {factory->name(), factory->createFilterFactoryFromProto(message, factory_context_)};
```

Comment:
As opposing to network filters, it seems like HTTP filters factory also takes the ``stats_prefix`` as a parameter. This doesn't effect the ECDS metrics, but I was wondering why network filter factories don't use this param?

Reply:
It shouldn't be added, it's from an old, mistaken way that used listener stat prefix. I think ECDS needs to have its own stat root, since it operates independently.